### ISSUE TITLE
Improve handling of connection failures

### ISSUE THREAD

**Issue by srh** [dgraph-io/dgraph#1121]
Right now we don't have a way to track the health of a peer and our connectivity towards it, or use that information.

We want the following:
 - To be able to respond to the health of a peer, instead of (for example) sitting around sending queries to a dead peer.  To track our node's opinion of peer reachability.
 - ~Edit: Adding: Use more than just notions of network availability for querying -- we have to wait for a peer to be up-to-date.~  Update: This is handled by #1344.
 - ~To tighten up the management of connection pools (don't keep around the old connection pool when an address has moved).~  (edit: this is done)

~Also, a microbug: `syncMemberships` uses `defer` in a loop right now.~ (edit: this is done)

**Comment by srh** [comment]
According to info online, the gRPC connection object, `grpc.ClientConn`, can be shared among goroutines.  We need to take advantage of this.

One big problem with our connection pool based approach is that it only performs one gRPC call at a time on a given TCP connection.  We also only have 5 spare connections.  That means individual RPC's will often suffer connection handshake latency, and it also means we might run out of file descriptors under high load.  (That would be the case unless query evaluation logic batches operations somehow.)

So, it looks like the right thing to do is:
- Drop connection pooling, and use a single `grpc.ClientConn` concurrently.
- If having only one becomes a problem, have more `grpc.ClientConn`'s that are used concurrently.

I'm going to try having just one `grpc.ClientConn` now.

**Comment by manishrjain** [comment]
Sounds good. Just check how many concurrent requests can be done over just one connection. Is there a limit? Would requests suffer from having just one connection? At what throughput, would you want to have more than 1 connection?

**Comment by srh** [comment]
There is no intrinsic limit to the number of concurrent requests, except there is a limit to the total number of requests per gRPC stream.  The reason is, it uses HTTP 2, which limits each end to making 2^30 streams over the lifetime of the connection.  Stream numbers can't repeat, so, you run out of them, and a new connection has to be made.

One sort of pain from having just one connection is when you get TCP packet loss, and then suddenly all the messages are backed up.  Whereas if you could use multiple streams or react to latency somehow (or use SCTP) you'd avoid that.  We might get single-digit numbers of users complaining about that, see https://github.com/rethinkdb/rethinkdb/issues/5852 for example.

I think you don't get more throughput from more streams unless there's a lot of packet loss or maybe (I don't know) if things are relatively exotic, like having more than one network interface.

**Comment by srh** [comment]
I'm currently thinking of ways to deal with the situation where both `node` and `groupi` keep track of a set of `(groupID uint32, nodeID uint64, addr string)` tuples (in `servers` and `peerPool` types).  The code seems relatively liberal about when and how the node's `peerPool` gets updated -- I have a feeling there's trouble there, where multiple things set the variable.  This is related to question of removing `poolsi` entries to unused connections.  I'll need to go over the code again tomorrow morning, but there's a good chance I'll just track peer health in `poolsi` and use reference counting there.

**Comment by manishrjain** [comment]
> There is no intrinsic limit to the number of concurrent requests,

We discussed this but just wanted to jot it here as well: Grpc by default only allows 100 concurrent requests. We can modify that to something higher (Dgraph client sets it to 1000), and the server is now set with unlimited concurrent streams per connection. Not sure what the trade off here is though.

**Comment by srh** [comment]
I think the server still has the default value, 100 per connection, because I see nothing setting MaxConcurrentStreams, just by grepping the codebase. For intracluster connections in branch sam/1121 (which replaces the connection pool with one connection in each direction) the setting will be unlimited.

**Comment by srh** [comment]
You don't actually want to use unlimited streams, it's just the place for that sort of resource management for cluster stuff is in the application.  For example, suppose you did want to limit the amount of concurrent streams.  What you want then is to do the accounting up front, at least approximately, and push back on incoming operations (that will create RPC's) right away, instead of waiting until you begin evaluating the query and then deciding you're using too many RPC resources once you kick off an RPC.  The reason being delayed feedback can create traffic waves under high load and limit throughout.

A stream is sort of lightweight -- in Go I think to wait on a stream you do have to block an entire goroutine though (I mean 8K of memory) -- the API doesn't expose a chan.

**Comment by manishrjain** [comment]
Agreed. But, I think the Grpc setting is just per connection. We'd want to do throttling globally on all connections; so if we're sitting idle, and some connections sends a lot of requests (like the loader does), the server shouldn't really reject. I think the Grpc stream limit feature (http/2 feature) doesn't really achieve that.

You're right about the blocking goroutine, would be great to avoid that memory cost.

**Comment by srh** [comment]
We want to include more than network connectivity when evaluating server health.  For example, we don't want to use a server that has just started up, or had recently been disconnected, and is retrieving a snapshot.  For example, right now we have the bug that if you run a query on a server that's retrieving a snapshot, the query will hang.  The reason is, it tries to run locally.  The query shouldn't hang (or run locally), instead it should be routed to a different server.  (I believe we'll also hang if we add a node that doesn't serve that group, and try to run queries from it -- one will get routed to the node that is retrieving a snapshot.)

I don't at this time know what mechanism specifically causes the query to hang instead of return stale data (I haven't looked), maybe somebody else does.

So we definitely want the following information:
- did the last attempt to communicate to the node succeed
- for each group the node is a member of, is the node ready to serve queries for that group (as opposed to, say, retrieving a snapshot)
- perhaps other information

**Comment by manishrjain** [comment]
We have a health checker which could cover the snapshot retrieval case (mark the server unhealthy). Also, if you implement read linearizability, the snapshot retrieval case would be achieved anyways.

I think right now it should just reject the query, instead of hanging it. If it hangs, look deeper and find why. @janardhan1993 can help. However, hanging a query isn't a bad outcome, not something we need to try to avoid. We just need mechanisms to retry on another server.

In any case, the job of redirect queries is not for the receiving server. It should lie with the caller. If we implement what Jeff Dean's method from his talk (think I shared with you), then the second query should run OK and ensure that our latency isn't affected.

**Comment by manishrjain** [comment]
What's required here? Should we block v0.8.1 on this?

**Comment by srh** [comment]
We should not block v0.8.1 on this.  We don't track health of peers yet.

**Comment by srh** [comment]
Anyway, I'm working on the second part of this issue now.  First doing the rapid response times technique, then adding some knowledge of server health so that we query healthy servers.

**Comment by srh** [comment]
One problem with doing the rapid-response time latency-reduction stuff is that we need to have a notion of when we "start" processing a query.  In other words, the number of actively processed requests has to be limited somehow.

We don't do that right now.  The Sort or ServeTask rpc call arrives, and we start processing it, without consideration for the number of concurrent queries being processed.  We can't implement cross-server request cancellation unless we have that.

**Comment by manishrjain** [comment]
We had that in place when we were seeing memory issues, but removed that since. In general it makes sense to have that, but what are the specific reasons for having it?

@janardhan1993 has more context regarding the rate limiter for queries.

**Comment by srh** [comment]
We shouldn't come up with one limit.  Throttling requests should be based on utilization of other resources that are the limit.  For example, suppose we have a memory-only system, processing requests.  (No disk reads.)  In that case, simplistically, there would be no point in accepting more requests if all CPUs are being utilized.  (We can get more nuanced by having prioritization between long-running requests and short-running requests.)  When we have a system where some queries talk to disk, you get to a point where there is no benefit in having more concurrent requests to disk.

Part of the problem I am trying to solve is: What happens if a machine gets requests faster than it can process? That is the situation that Jeff Dean's rapid response times talk is about.  If that situation is out of scope, there is no reason to implement cross-server request cancellation.  But also, the design of cross-server request cancellation requires that requests be blocked in some queue in the situation that a server is overloaded.

**Comment by manishrjain** [comment]
The general issue here is that it's hard to know what to limit by. We have mechanisms in place to drop requests if our memory usage exceeds a certain pre-specified threshold. We can probably put something in place for CPU usage as well, maybe.

In Google systems, all the requests are nicely queued up. That's not exposed in Grpc. But, the general idea still stands, is that we can cancel a context, which would then translate to canceling that request whatever stage it is at. Jeff talked about doing this upfront. For our situation, it might be simpler to do it when a query finishes (as opposed to be starts) processing. With the time delay of 2ms, this would still be a good start for us. And if we introduce message queue, we can swap this.

We also have a limit for pending proposals. I think when you implement read linearizability, you can rate limit the number of pending read requests which can then be the throttled, if required.

**Comment by janardhan1993** [comment]
@srh Can this be closed ?

**Comment by srh** [comment]
We don't track our node's opinion of peer reachability yet.

**Comment by manishrjain** [comment]
Fixed with https://github.com/dgraph-io/dgraph/commit/dc8314de8c239daf84ce1541d02a54d4ac4dea6f
